crates/biome_js_analyze/src/lint/nursery/no_restricted_types.rs

```diff
@@ -0,0 +1,356 @@
+                let token_name = identifier_token.text_trimmed();
+
+                let restricted_type =
+                    if should_extend_defaults && model.binding(reference_identifier).is_none() {
+                        RestrictedType::from_str(token_name)
+                    } else {
+                        options.types.get(token_name).map(|custom_restricted_type| {
+                            RestrictedType::Custom((
+                                token_name.to_string(),
+                                custom_restricted_type.clone(),
+                            ))
+                        })
+                    }?;
+
+                return Some(State {
+                    restricted_type,
+                    range: identifier_token.text_trimmed_range(),
+                    reference_identifier: Some(reference_identifier.clone()),
+                });
+            }
+        }
+
+        None
+    }
+
+    fn diagnostic(_ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let restricted_type = &state.restricted_type;
+
+        let message = match restricted_type {
+            RestrictedType::Custom((_, data)) => &data.message,
+            _ => &format!("Don't use '{}' as a type.", restricted_type),
+        };
+
+        let mut diagnostic = RuleDiagnostic::new(
+            rule_category!(),
+            state.range,
+            markup! { {message} }.to_owned(),
+        );
+
+        if let Some(additional_note) = restricted_type.additional_note() {
+            diagnostic = diagnostic.note(markup! { {additional_note} }.to_owned());
+        }
+
+        Some(diagnostic)
+    }
+
+    fn action(ctx: &RuleContext<Self>, state: &Self::State) -> Option<JsRuleAction> {
+        let mut mutation = ctx.root().begin();
+        let suggested_type = state.restricted_type.suggested_fix_type()?;
+
+        let prev_token = state.reference_identifier.clone()?.value_token().ok()?;
+        let new_token = make::ident(suggested_type.as_str());
+
+        mutation.replace_element(prev_token.into(), new_token.into());
+
+        Some(JsRuleAction::new(
+            ActionCategory::QuickFix,
+            ctx.metadata().applicability(),
+            markup! { "Use '"{suggested_type}"' instead" }.to_owned(),
+            mutation,
+        ))
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserializable, Deserialize, Serialize, Eq, PartialEq)]
+#[cfg_attr(feature = "schemars", derive(JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct NoRestrictedTypesOptions {
+    extend_defaults: Option<bool>,
+    types: FxHashMap<String, CustomRestrictedType>,
+}
+
+#[derive(Debug, Clone, Default, Deserializable, Deserialize, Serialize, Eq, PartialEq)]
+#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct CustomRestrictedType {
+    message: String,
+    fix_with: Option<String>,
```

Comment:
We could make the message optional and generate a default message such as `Use <fix_with> type instead`.
Also, I could rename `fix_with` with `use`.
```suggestion
    message: Option<String>,
    use: Option<String>,
```

We could add a validation state that ensures that `use` is a valid identifier (a function `js_ident` allows checking it) and at least one field is specified (either `message`, or `use`, or both).

Reply:
I renamed the field, not sure about making `message` field optional, banning types should be very intentional action and default explanation would not be helpful for other people. 

I would like to make option to accept string directly like this
```
type Options = {
  foo: string | { message: string }
}
```
Though I am not sure how to do that with Serde